comtypes/_memberspec.py

```diff
@@ -173,62 +173,97 @@ def _fix_inout_args(
-                        # call the constructor with the argument
-                        # in that case.
+                        # The from_param method of simple types (c_int, c_double, ...) returns a byref()
+                        # object which we cannot use since later it will be wrapped in a pointer.  Simply
+                        # call the constructor with the argument in that case.
                         v = atyp(v)
                     else:
                         v = atyp.from_param(v)
                         assert not isinstance(v, BYREFTYPE)
-                outargs[outnum] = v
-                outnum += 1
-                if len(args) > i:
-                    args[i] = v
-                else:
+                    return v
+
+                if is_positional:
+                    v = prepare_parameter(args[param_index])
+                    args[param_index] = v
+                elif name in kw:
+                    v = prepare_parameter(kw[name])
                     kw[name] = v
-            elif direction & 2 == 2:
+                else:
+                    # no parameter was passed, make an empty one of the required type
+                    # and pass it as a keyword argument
+                    v = atyp()
+                    if name is not None:
+                        kw[name] = v
+                    else:
+                        raise Exception("Unnamed inout parameters cannot be omitted")
+                outargs[outnum] = v
+            if dir_out:
                 outnum += 1
+            if dir_in:
+                param_index += 1
+
         rescode = func(self, *args, **kw)
         # If there is only a single output value, then do not expect it to
         # be iterable.
+
+        # My interpretation of this code:
+        # - 'outnum' counts the total number of 'out' and 'inout' arguments.
+        # - Confusingly, 'outargs' is a dict consisting of the supplied _inout_ arguments.
+        # - The call to func() returns the 'out' and 'inout' arguments.
+        #   Furthermore, it changes the variables in 'outargs' as a "side effect"
+        # - In a perfect world, it should be fine to just return 'rescode', but I assume there is a reason
+        #   why the original authors did not do that. Instead, they replace the inout variables in 'rescode'
+        #   by those in 'outargs', and call __ctypes_from_outparam__() on them.
+        #
+        # - What I don't understand is: Why is the behaviour different for outnum==1?
+        #   From the above interpretation I would have expected
+        #
+        #   rescode = outargs[0].__ctypes_from_outparam__()
+        #
+        #   instead of
+        #
+        #   rescode = rescode.__ctypes_from_outparam__()
+        #
+        #
+
         if outnum == 1:  # rescode is not iterable
             if len(outargs) == 1:
                 rescode = rescode.__ctypes_from_outparam__()
+                # rescode = outargs[0].__ctypes_from_outparam__()
```

Comment:
I think this comment is not necessary since such line is mentioned in the above large comment.